### PR TITLE
e2e: upgrade: tests for topology manager

### PR DESCRIPTION
**Pull request by ffromani**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This is another followup of #86184 . I'm working with @vpickard to add more e2e tests for the topology manager to fix the issue #83481

**Which issue(s) this PR fixes**:
Related to #83481

**Special notes for your reviewer**:
N/A

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by fromanirh** [comment]
/hold
still WIP

File: test/e2e/upgrades/topology_manager.go
Comment (fromanirh) [inline test/e2e/upgrades/topology_manager.go]: I have another PR in the working to fix this

Comment (fromanirh) [reply test/e2e/upgrades/topology_manager.go]: Here: https://github.com/kubernetes/kubernetes/pull/88110

File: test/e2e/upgrades/topology_manager.go
Comment (oomichi) [inline test/e2e/upgrades/topology_manager.go]: ditto

File: test/e2e/upgrades/topology_manager.go
Comment (oomichi) [inline test/e2e/upgrades/topology_manager.go]: A necessary blank line between `"k8s.io/*"` block except `"k8s.io/kubernetes"` and `"k8s.io/kubernetes"` block.

File: test/e2e/upgrades/topology_manager.go
Comment (oomichi) [inline test/e2e/upgrades/topology_manager.go]: ditto: unnecessary blank line.

File: test/e2e/upgrades/topology_manager.go
Comment (oomichi) [inline test/e2e/upgrades/topology_manager.go]: ditto: unnecessary blank line.

File: test/e2e/upgrades/topology_manager.go
Comment (oomichi) [inline test/e2e/upgrades/topology_manager.go]: ditto: unnecessary blank line.

File: test/e2e/upgrades/topology_manager.go
Comment (oomichi) [inline test/e2e/upgrades/topology_manager.go]: What is this `XXX`? `NOTE` or something?

File: test/e2e/upgrades/topology_manager.go
Comment (oomichi) [inline test/e2e/upgrades/topology_manager.go]: nit: better to put a new blank line before `ginkgo.By()` for code readability.

File: test/e2e/upgrades/topology_manager.go
Comment (fromanirh) [reply test/e2e/upgrades/topology_manager.go]: I missed how can I check for these style violation on my box, saving review time. Could you please suggest a way (o r point me to the right docs :)) ?

File: test/e2e/upgrades/topology_manager.go
Comment (fromanirh) [reply test/e2e/upgrades/topology_manager.go]: was a `NOTE` to self which in retrospect is just misleading. Removed.

File: test/e2e/upgrades/topology_manager.go
Comment (fromanirh) [reply test/e2e/upgrades/topology_manager.go]: done

File: test/e2e/upgrades/topology_manager.go
Comment (oomichi) [inline test/e2e/upgrades/topology_manager.go]: Can I know why this pod requires 2 CPU cores?
That can be too much on some environments, I feel.

File: test/e2e/upgrades/topology_manager.go
Comment (oomichi) [inline test/e2e/upgrades/topology_manager.go]: Does this require that all CPU cores should be in the same siblings from a pod viewpoint?
BTW typo: s/sibilings/siblings/

**Review by oomichi — COMMENTED** [review]
Thanks for updating.
Let's test anyways.

/ok-to-test

File: test/e2e/upgrades/topology_manager.go
Comment (fromanirh) [reply test/e2e/upgrades/topology_manager.go]: Two cores is the minimum amount of resources we can use to actually check the resources are aligned. But here we want to check if the topology manager works correctly across upgrades, hence we maybe can simplify a bit more and require just one core. I'll do some research and put the PR on hold until then.

File: test/e2e/upgrades/topology_manager.go
Comment (fromanirh) [reply test/e2e/upgrades/topology_manager.go]: fixed typo and clarified the intent of the test.

**Comment by fromanirh** [comment]
/hold
need to see if we can reduce the resource request to just 1 core

File: test/e2e/upgrades/topology_manager.go
Comment (fromanirh) [reply test/e2e/upgrades/topology_manager.go]: The topology manager is actually exercised using just one core, but I can't find a way to test it worked as intended with less than two cores, so I need to keep it this way.

**Comment by fromanirh** [comment]
/hold cancel
it seems two full cores is really the minimum we need to test the topology manager is working

File: test/e2e/upgrades/topology_manager.go
Comment (derekwaynecarr) [reply test/e2e/upgrades/topology_manager.go]: I think its fine that environments which require alignment testing verification during upgrades can reserve 2 cores.

**Review by derekwaynecarr — APPROVED** [review]
this looks good from sig-node perspective.

/approve
/lgtm

**Comment by oomichi** [comment]
Thanks for updating.

/lgtm

**Comment by cynepco3hahue** [comment]
@derekwaynecarr @oomichi Hi guys, I can see that we skip upgrade tests under the `pull-kubernetes-e2e-gce`, maybe do we have some CI lane that can be initiated manually to run upgrade tests?

**Comment by fromanirh** [comment]
I tried to run the e2e upgrade flow on my box on gcp with no success yet, do we have up to date instructions?

**Comment by fromanirh** [comment]
> I tried to run the e2e upgrade flow on my box on gcp with no success yet, do we have up to date instructions?

Looks like it works now: I was just lacking a couple of gsutil/gcloud commands in the steps I followed. I'll update as soon as I have some progress.

**Comment by fromanirh** [comment]
how come the approved label wasn't removed? I think we need another review, the last changes are pretty significant.

**Comment by oomichi** [comment]
@fromanirh 
You need to update this PR with `./hack/update-bazel.sh` to pass pull-kubernetes-verify test job.

File: test/e2e/cloud/gcp/cluster_upgrade.go
Comment (oomichi) [inline test/e2e/cloud/gcp/cluster_upgrade.go]: At this time, test-infra contains just `"\[Feature:TopologyManager\]"` without `Upgrade` keyword in sig-node/node-kubelet.yaml and sig-node/sig-node-presubmit.yaml.
Will you add the above keyword also? Or should we change here to `"\[Feature:TopologyManager\]"` to run this test on the above test jobs?

**Review by oomichi — COMMENTED** [review]
@cynepco3hahue 

> @derekwaynecarr @oomichi Hi guys, I can see that we skip upgrade tests under the pull-kubernetes-e2e-gce, maybe do we have some CI lane that can be initiated manually to run upgrade tests?

That is a nice point.
The test contains `[Feature:***]` label which tests are skipped in general test jobs as  https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/88107/pull-kubernetes-e2e-gce/1233340232040976384/build-log.txt which shows `--ginkgo.skip=\[Slow\]|\[Serial\]|\[Disruptive\]|\[Flaky\]|\[Feature:.+\]` is specified.
We can see the test result on different test jobs which are specific SIG ones.

According to the above comment, I have a question about this test job.
Please check my inline comment.

**Comment by fromanirh** [comment]
@oomichi ack, I've seen your question but I need to sync in order to provide an answer. Will respond ASAP.

Meanwhile: I updated the tests and squashed a few little bugs in the process. Yet, they still don't complete on my env. However also a clean master snapshot doesn't complete for me. Here's what I did:

k8s master
	fdb2cb4c883
build using
```kubetest --build --stage=gs://ABC --check-version-skew=false --gcp-project=XXX```
INTENTIONALLY made all objects in the bucket world-readable (AFAIU the e2e tests require this, but this is just a nightly build so nothing to hide):
```gsutil -m acl -r ch -u AllUsers:R gs://ABC```
run using
``kubetest --up --test --check-version-skew=false --extract=gs://ABC/v1.19.0-alpha.0.399+fdb2cb4c8832da --extract=v1.17.2 --upgrade_args=--ginkgo.focus="Feature:Upgrade" --gcp-project=XXX --down```

I narrowed down to this: when the test run cluster/gce/upgrade.sh, the test get stuck there

```
  echo "== Waiting for new node to be added to k8s.  ==" >&2
  while true; do
    local new_boot_id=$("${KUBE_ROOT}/cluster/kubectl.sh" get node "${instance}" --output=jsonpath='{.status.nodeInfo.bootID}' 2>&1) && kubectl_rc=$? || kubectl_rc=$?
    if [[ "${kubectl_rc}" != 0 ]]; then
      echo "== FAILED to get node ${instance} =="
      echo "${boot_id}"
      echo "  (Will retry.)"
    elif [[ "${boot_id}" != "${new_boot_id}" ]]; then
      echo "Node ${instance} recreated."
      break
    else
      echo -n .
    fi
    sleep 1
  done
```
it seems the `bootID` never changes. I waited for 30m with no changes. Maybe I just need to wait more? 

Any suggestion to make it move forward is welcome.

**Comment by fromanirh** [comment]
quick update: I had to work on other stuff, but I definitely wants to understand what's wrong there and I should be able to work on this again soon. Thanks.

**Comment by oomichi** [comment]
@fromanirh 

> @oomichi ack, I've seen your question but I need to sync in order to provide an answer. Will respond ASAP.

I am just waiting for your answer for the above.

File: test/e2e/cloud/gcp/cluster_upgrade.go
Comment (fromanirh) [reply test/e2e/cloud/gcp/cluster_upgrade.go]: Will sync with the node team and figure out, I don't have an answer for this.

**Comment by fromanirh** [comment]
> @fromanirh
> 
> > @oomichi ack, I've seen your question but I need to sync in order to provide an answer. Will respond ASAP.
> 
> I am just waiting for your answer for the above.

Absolutely, sorry for the delay. I don't have yet an answer, need to get in touch with relevant people.  Hopefully this will happen next week. 
What I meant in my previous comment is that I still have some investigation to do on my side (hence the PR still on hold)

File: test/e2e/cloud/gcp/cluster_upgrade.go
Comment (fromanirh) [reply test/e2e/cloud/gcp/cluster_upgrade.go:286]: Hi, partial followup, still figuring out the path forward with the other parties.
I was trying to make my code work and eventually checked out how the existing code is working, but according to https://testgrid.k8s.io/google-gce-upgrade even the existing tests are red, and the last upgrade check configured was from 1.13 and 1.14. This was unexpected, am I looking in the wrong place perhaps? If so what's the right place to check how existing test are doing?
If the existing tests are indeed red the simplest way forward is drop this PR and restart from scratch in the node area.
Thanks for any help.

**Comment by fromanirh** [comment]
considering the work begun on sig-node to revamp and clean up our e2e tests, I'm closing this PR. We'll evaluate how to introduce a new upgrade test there.
Thanks to all the reviewers.

**Pull request closed without merging**
